src/pip/_internal/req/req_install.py

```diff
@@ -597,24 +602,59 @@ def prepare_metadata(self) -> None:
 
     @property
     def metadata(self) -> Any:
+        # TODO: use cached_property in python 3.8+
         if not hasattr(self, "_metadata"):
-            self._metadata = self.get_dist().metadata
+            self._metadata = self.cached_dist.metadata
 
         return self._metadata
 
-    def get_dist(self) -> BaseDistribution:
-        if self.metadata_directory:
-            return get_directory_distribution(self.metadata_directory)
-        elif self.local_file_path and self.is_wheel:
-            assert self.req is not None
-            return get_wheel_distribution(
-                FilesystemWheel(self.local_file_path),
-                canonicalize_name(self.req.name),
+    @property
+    def cached_dist(self) -> BaseDistribution:
+        """Retrieve the dist resolved from this requirement.
+
+        :raises AssertionError: if the resolver has not yet been executed.
+        """
+        if self._cached_dist is None:
+            raise AssertionError(
+                f"InstallRequirement {self} has no dist; "
+                "ensure the resolver has been executed"
             )
-        raise AssertionError(
-            f"InstallRequirement {self} has no metadata directory and no wheel: "
-            f"can't make a distribution."
-        )
+        return self._cached_dist
+
+    def cache_virtual_metadata_only_dist(self, dist: BaseDistribution) -> None:
+        """Associate a "virtual" metadata-only dist to this requirement.
+
+        This dist cannot be installed, but it can be used to complete the resolve
+        process.
+
+        :raises AssertionError: if a dist has already been associated.
+        :raises AssertionError: if the provided dist is "concrete", i.e. exists
+                                somewhere on the filesystem.
+        """
+        assert self._cached_dist is None, self
+        assert not dist.is_concrete, dist
+        self._cached_dist = dist
+
+    def cache_concrete_dist(self, dist: BaseDistribution) -> None:
+        """Associate a "concrete" dist to this requirement.
+
+        A concrete dist exists somewhere on the filesystem and can be installed.
+
+        :raises AssertionError: if a concrete dist has already been associated.
+        :raises AssertionError: if the provided dist is not concrete.
+        """
+        if self._cached_dist is not None:
+            # If we set a dist twice for the same requirement, we must be hydrating
+            # a concrete dist for what was previously virtual. This will occur in the
+            # case of `install --dry-run` when PEP 658 metadata is available.
+            if not self.allow_concrete_dist_overwrite:
+                assert not self._cached_dist.is_concrete
```

Comment:
I _really_ don't like the idea of having behaviours change based on whether we're under a specific test case or not, like this.

Have you considered alternatives that do not require such logic-only-changed-for-testing, or otherwise mocking things out in the test suite so that the behaviour change is more clearly contained within the tests themselves?

This matters a bit more since we don't actually exercise code coverage in this codebase, so it's fairly easy to miss/overlook removal of this logic if the test evolves.

Reply:
I have fixed this by using `pytest.MonkeyPatch` instead! Thank you so much for elaborating on this point!